File: apps/weather-frontend/client/src/services/weatherService.ts
import type { WeatherAPIResponse, WeatherInfo } from '@/types/weather';

const API_URL = 'https://api-open.data.gov.sg/v2/real-time/api/four-day-outlook';

export async function fetchWeatherData(): Promise<WeatherInfo | null> {
  try {
    const response = await fetch(API_URL, {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      console.error('API Request Failed. Response Code:', response.status);
      return null;
    }

    const data: WeatherAPIResponse = await response.json();
    return parseWeatherData(data);
  } catch (error) {
    console.error('Error fetching weather data:', error);
    return null;
  }
}

function parseWeatherData(apiResponse: WeatherAPIResponse): WeatherInfo | null {
  try {
    const records = apiResponse.data.records;
    
    if (records.length === 0) {
      return null;
    }

    const record = records[0];
    const weatherInfo: WeatherInfo = {
      location: 'Singapore',
      timestamp: record.timestamp,
      forecasts: record.forecasts.map(forecast => ({
        date: forecast.day,
        weather: `${forecast.forecast.text} - ${forecast.forecast.summary}`,
        temperature: {
          low: forecast.temperature.low,
          high: forecast.temperature.high,
        },
        humidity: {
          low: forecast.relativeHumidity.low,
          high: forecast.relativeHumidity.high,
        },
        wind: {
          speedLow: forecast.wind.speed.low,
          speedHigh: forecast.wind.speed.high,
          direction: forecast.wind.direction,
        },
      })),
    };

    return weatherInfo;
  } catch (error) {
    console.error('Error parsing weather data:', error);
    return null;
  }
}

// 辅助函数
export function getWeatherAdvice(weather: string): string {
  const lowerWeather = weather.toLowerCase();
  if (lowerWeather.includes('shower') || lowerWeather.includes('rain')) {
    return '☔ Rain expected, carry umbrella and wear waterproof shoes';
  } else if (lowerWeather.includes('fair') || lowerWeather.includes('sunny')) {
    return '☀️ Good weather for outdoor sightseeing, don\'t forget sunscreen';
  } else if (lowerWeather.includes('cloud')) {
    return '⛅ Cloudy but pleasant, good for walking tours';
  } else if (lowerWeather.includes('wind')) {
    return '💨 Windy conditions, secure loose items';
  }
  return 'Enjoy your day!';
}

export function getWindAdvice(windSpeed: number): string {
  if (windSpeed < 10) return 'Calm conditions, perfect for outdoor activities';
  if (windSpeed < 20) return 'Light breeze, good for flying kites';
  if (windSpeed < 30) return 'Moderate wind, be careful with umbrellas';
  return 'Windy conditions, consider indoor activities';
}

export function getHumidityAdvice(humidity: number): string {
  if (humidity < 40) return 'Low humidity, remember to drink water';
  if (humidity < 70) return 'Comfortable humidity level';
  if (humidity < 85) return 'High humidity, may feel sticky';
  return 'Very high humidity, avoid strenuous outdoor activities';
}

export function formatTimestamp(timestamp: string): string {
  return timestamp.replace('+08:00', '').replace('T', ' ');
}

File: apps/weather-api/src/weather/WeatherService.java
package weather;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import org.json.JSONArray;
import org.json.JSONObject;

public class WeatherService {
    private static final String API_URL = "https://api-open.data.gov.sg/v2/real-time/api/four-day-outlook";

    public WeatherInfo fetchWeatherData() {
        WeatherInfo weatherInfo = null;
        
        try {
            // 1. 建立连接
            URL url = new URL(API_URL);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("Accept", "application/json");

            // 2. 检查响应状态
            int responseCode = connection.getResponseCode();
            if (responseCode != 200) {
                System.out.println("API Request Failed. Response Code: " + responseCode);
                return null;
            }

            // 3. 读取响应
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            // 4. 解析JSON数据
            weatherInfo = parseWeatherData(response.toString());
            
        } catch (Exception e) {
            System.err.println("Error fetching weather data: " + e.getMessage());
            e.printStackTrace();
        }
        
        return weatherInfo;
    }

    private WeatherInfo parseWeatherData(String jsonResponse) {
        try {
            JSONObject json = new JSONObject(jsonResponse);
            
            // 根据你提供的实际JSON结构进行解析
            JSONObject data = json.getJSONObject("data");
            JSONArray records = data.getJSONArray("records");
            
            if (records.length() == 0) {
                return null;
            }
            
            // 获取第一条记录（应该只有一条）
            JSONObject record = records.getJSONObject(0);
            String timestamp = record.getString("timestamp");
            String date = record.getString("date");
            
            // 创建WeatherInfo对象
            WeatherInfo weatherInfo = new WeatherInfo("Singapore", timestamp);
            
            // 解析天气预报数据
            JSONArray forecasts = record.getJSONArray("forecasts");
         // 在解析循环中添加湿度和风速的解析
            for (int i = 0; i < forecasts.length(); i++) {
                JSONObject forecast = forecasts.getJSONObject(i);
                
                String day = forecast.getString("day");
                String weatherText = forecast.getJSONObject("forecast").getString("text");
                String weatherSummary = forecast.getJSONObject("forecast").getString("summary");
                
                // 解析温度
                JSONObject temp = forecast.getJSONObject("temperature");
                int low = temp.getInt("low");
                int high = temp.getInt("high");
                WeatherInfo.Temperature temperature = new WeatherInfo.Temperature(low, high);
                
                // 解析湿度（新增）
                JSONObject humidity = forecast.getJSONObject("relativeHumidity");
                int humidityLow = humidity.getInt("low");
                int humidityHigh = humidity.getInt("high");
                WeatherInfo.Humidity humidityObj = new WeatherInfo.Humidity(humidityLow, humidityHigh);
                
                // 解析风速和风向（新增）
                JSONObject wind = forecast.getJSONObject("wind");
                JSONObject windSpeed = wind.getJSONObject("speed");
                int windSpeedLow = windSpeed.getInt("low");
                int windSpeedHigh = windSpeed.getInt("high");
                String windDirection = wind.getString("direction");
                WeatherInfo.Wind windObj = new WeatherInfo.Wind(windSpeedLow, windSpeedHigh, windDirection);
                
                // 创建单日预报（更新构造函数）
                WeatherInfo.DayForecast dayForecast = 
                    new WeatherInfo.DayForecast(day, weatherText + " - " + weatherSummary, 
                                               temperature, humidityObj, windObj);
                weatherInfo.addForecast(dayForecast);
            }
                
                
            
            
            return weatherInfo;
            
        } catch (Exception e) {
            System.err.println("Error parsing weather data: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }
}